### ISSUE TITLE
Support for RBAC access to Cosmos account

### ISSUE THREAD

**Issue by ilya-git** [terraform-providers/terraform-provider-azurerm#13907]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritise this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritise the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Azure has a new feature that allows to access data in Cosmos Account using RBAC. Ordinary RBAC assignment does not work there, so it's something that is specific to Cosmos. I did not find a way to do this in terraform right now.

RBAC allows for more fine-grained control by using various actions, so it would be nice to have a possibility to use these as well, but they can perhaps be used via a custom role. Anyway just the role assignment itself will already be a very nice feature.

### New or Affected Resource(s)

* azuread_cosmos_account_role_assignment

### Potential Terraform Configuration

```hcl
resource "azuread_cosmos_account_role_assignment" "mydb" {
  service_principal_object_id = <object_id>
  cosmos_account_name  = <account_name>
  cosmos_account_resource_group = <resource_group_name>
  role_definition_name = "Cosmos DB Built-in Data Contributor"
}
```

### References

* Here are more details about RBAC access: https://docs.microsoft.com/en-us/azure/cosmos-db/how-to-setup-rbac

**Comment by adamyager** [comment]
This does now seem unblocked correct? Any assistance one getting this in the backlog would be appreciated.

**Comment by aristosvo** [comment]
[Upstream issue](https://github.com/Azure/azure-rest-api-specs/issues/16560) is still open unfortunately, the issue on the Go implementation is closed as upstream should fix it.

**Comment by adamyager** [comment]
> [Upstream issue](https://github.com/Azure/azure-rest-api-specs/issues/16560) is still open unfortunately, the issue on the Go implementation is closed as upstream should fix it.

Ah, appreciate the details. Thanks!

**Comment by gsadhana** [comment]
Any update on this?

**Comment by adamyager** [comment]
@aristosvo  this was just closed. Good news. 
Just an FYI

**Comment by Carike** [comment]
Hi everyone,

Just checking whether the change has been merged or still requires additional testing?

Thanks

**Comment by Carike** [comment]
Hi @aristosvo & @manicminer,

Trust you are all well.

Can the fix (https://github.com/aristosvo/terraform-provider-azurerm/tree/cosmosdb/rbac-role-assignment ) be merged since the upstream issue has been resolved? 

Would really appreciate it, as this is currently a blocker on one of our tasks. Thank you again!

**Comment by aristosvo** [comment]
I believe this is already merged, but based on a different implementation. Let me check and circle back with more info.

**Comment by ilya-git** [comment]
Is there a way to use the built-in roles? I did not find any explicit documentation on how to do this:

This is what I tried to "reference" the role:

```
resource "azurerm_cosmosdb_sql_role_definition" "example" {
  resource_group_name = local.cosmos_resource_group_name
  account_name        = local.cosmos_account_name
  name                = local.cosmos_role_definition_name
  type                = BuiltInRole
}
```

But I get this error: 

> A reference to a resource type must be followed by at least one attribute access, specifying the resource name.

**Comment by arkiaconsulting** [comment]
@ilya-git missing quotes ? => "BuiltInRole"

**Comment by ilya-git** [comment]
Yes, thank you, I have found this actually as well and this is what I ended up doing:

```
resource "azurerm_cosmosdb_sql_role_definition" "example" {
  role_definition_id  = local.cosmos_role_definition_id
  resource_group_name = local.cosmos_resource_group_name
  account_name        = local.cosmos_account_name
  name                = "Cosmos DB Built-in Data Contributor"
  assignable_scopes   = []
  type                = "BuiltInRole"

  permissions {
    data_actions = []
  }
}
```

Got an error:

>  A resource with the ID ".../account-name/sqlRoleDefinitions/00000000-0000-0000-0000-000000000002" already exists - to be managed via Terraform this resource needs to be imported into the State. Please see the resource documentation for "azurerm_cosmosdb_sql_role_definition" for more information.

So I still cannot figure out how to use the BuiltInRole properly
